### ISSUE TITLE
Sending does not work

### ISSUE THREAD

**Issue by revoxhere** [stjet/nanohakase#4]
Hey, I'm having an issue gettin  the library to send XNO. It used to work, but since a few months it's returning a non-hexadecimal number error:

```python3
nrpc = nanoRPC("https://app.natrium.io/api")
nano_key = nanoWallet(nrpc, seed="XXXXX", index=0)
print(nano_key.get_balance())
print(nano_key.send("nano_3pdripjhteyymwjnaspc5nd96gyxgcdxcskiwwwoqxttnrncrxi974riid94", "0.001"))
```

```
{'balance': '2594130159959955286961', 'pending': '0', 'receivable': '0'}
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/__main__.py", line 7, in <module>
    run()
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/app/wsgiapp.py", line 67, in run
    WSGIApplication("%(prog)s [OPTIONS] [APP_MODULE]").run()
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/app/base.py", line 236, in run
    super().run()
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/app/base.py", line 72, in run
    Arbiter(self).run()
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/arbiter.py", line 58, in __init__
    self.setup(app)
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/arbiter.py", line 118, in setup
    self.app.wsgi()
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/app/wsgiapp.py", line 58, in load
    return self.load_wsgiapp()
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/app/wsgiapp.py", line 48, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/home/debian/.local/lib/python3.9/site-packages/gunicorn/util.py", line 371, in import_app
    mod = importlib.import_module(module)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/debian/duino-coin-master-server/app.py", line 562, in <module>
    print(nano_key.send("nano_3pdripjhteyymwjnaspc5nd96gyxgcdxcskiwwwoqxttnrncrxi974riid94", "0.001"))
  File "/home/debian/.local/lib/python3.9/site-packages/nanohakase/wallet.py", line 50, in send
    block_hash = hash_block(block)
  File "/home/debian/.local/lib/python3.9/site-packages/nanohakase/util.py", line 110, in hash_block
    blake_obj.update(hex_to_bytes(padded_balance))
  File "/home/debian/.local/lib/python3.9/site-packages/nanohakase/util.py", line 65, in hex_to_bytes
    return bytes.fromhex(hex)
ValueError: non-hexadecimal number found in fromhex() arg at position 8
```
I believe this is a bug in the library, as it used to work fine and my code hasn't changed. Can you maybe take a look at it in your free time?

**Comment by stjet** [comment]
Since how many months ago? And consistently? I just tried, worked fine for me: https://nanoblockexplorer.com/explorer/block/BC65BA179119FA354699CC49E01D2CF347D098B66D5C983A6196B581EC15AD51

Since I don't see any issues, if you could go into `/home/debian/.local/lib/python3.9/site-packages/nanohakase/util.py` and add a
```python
print(padded_balance, block)
```
around line 110 before the `blake_obj.update(hex_to_bytes(padded_balance))`? Then report back whatever it prints out? Thank you.

**Comment by revoxhere** [comment]
Thanks for your fast response. I believe it stopped working around December 2023.
I've added the print statement and here's the result:

```
{'balance': '2594130159959955286961', 'pending': '0', 'receivable': '0'}
00000000-33b2dafff06cc0b839ad44f {'type': 'state', 'account': 'nano_3uukauizf6xgypu6prfw55adg95k7iqpx7ttyu6onrtgcw66ech7kmridxu1', 'previous': '54C1C70F58DCB762BD26B5F1861DF157EDEF14BF17EF3E8BC010830CF53445DE', 'representative': 'nano_3uukauizf6xgypu6prfw55adg95k7iqpx7ttyu6onrtgcw66ech7kmridxu1', 'balance': '-999997405869840040044713039', 'link': 'D97885A2FD33DE9F234466CA1D16723BDD7297D56650E7395BF75AA628AC7607', 'link_as_account': 'nano_3pdripjhteyymwjnaspc5nd96gyxgcdxcskiwwwoqxttnrncrxi974riid94'}Traceback (most recent call last):
  File "/home/debian/duino-coin-master-server/btest.py", line 29, in <module>
    print(nano_key.send("nano_3pdripjhteyymwjnaspc5nd96gyxgcdxcskiwwwoqxttnrncrxi974riid94", "0.001"))
  File "/home/debian/.local/lib/python3.9/site-packages/nanohakase/wallet.py", line 50, in send
    block_hash = hash_block(block)
  File "/home/debian/.local/lib/python3.9/site-packages/nanohakase/util.py", line 112, in hash_block
    blake_obj.update(hex_to_bytes(padded_balance))
  File "/home/debian/.local/lib/python3.9/site-packages/nanohakase/util.py", line 65, in hex_to_bytes
    return bytes.fromhex(hex)
ValueError: non-hexadecimal number found in fromhex() arg at position 8
```


Side note: I believe it's working on a different server of mine, but the balance is all wrong (it displays only a very small fraction of available balance, while the key holds a lot more) and I can only send these small amonuts, but I think that may be a different issue, I'll have to check to be sure though.

**Comment by stjet** [comment]
Notice how the balance is negative: `'balance': '-999997405869840040044713039'`? Means you are sending more than your account has. The decimal to hex logic doesn't account for any negative numbers, so you can see the `00000000-33b2dafff06cc0b839ad44f` has a `-` in it. This problem has been solved for bananopie for a while, but I haven't updated the fork so the error message here is unclear - but anyways, just know that that error means you are sending more than you have. One day I might update nanohakase from the upstream bananopie but it's a bit of a pain.

**Comment by revoxhere** [comment]
Okay, but there's more than 500 XNO in the account I'm sending from, get_balance also shows a higher number than the amount I'm trying to send.

Here's a screenshot from the Exodus wallet that uses this (and only this) private key:
![image](https://github.com/stjet/nanohakase/assets/50244265/19aa1e80-1b26-4cba-bf4e-3ee645c59b88)
How's it possible that it shows negative, when I'm trying to send only a small fraction? I don't get it...

**Comment by stjet** [comment]
Weird - the logged block says the `account` is https://nanoblockexplorer.com/explorer/account/nano_3uukauizf6xgypu6prfw55adg95k7iqpx7ttyu6onrtgcw66ech7kmridxu1/history, which has <0.01 nano. I'm not sure how exodus works, but if it uses private keys as you say, be aware that private keys and seeds are different.

**Comment by jamescoxon** [comment]
Can I check that your are looking at the right accounts - 
the account @stjet references (https://nanoblockexplorer.com/explorer/account/nano_3uukauizf6xgypu6prfw55adg95k7iqpx7ttyu6onrtgcw66ech7kmridxu1/history) has received from the main account which @revoxhere is using (https://nanoblockexplorer.com/explorer/account/nano_3fpqpbcgt3nga3s81td6bk7zcqdr7ockgnyjkcy1s8nfn98df6c5wu14fuuq/history)

My suspicions is that either the seed isn't the right one or the index isn't correct (or using a private key instead of a seed).

**Comment by stjet** [comment]
The reason why I believe that it the account is `nano_3uukau....idxu1` due to the block that @revoxhere logged.. it has that address as its `account` (https://github.com/stjet/nanohakase/issues/4#issuecomment-2016421450). Additionally, the balance that revox logs, "2594130159959955286961" is much less than in the screenshot, and matches the balance for the 3uukau address (https://nanoblockexplorer.com/explorer/block/54C1C70F58DCB762BD26B5F1861DF157EDEF14BF17EF3E8BC010830CF53445DE).

@revoxhere , could you do `print(nano_key.get_address())`? That should make it clear what address is being used. Like @jamescoxon mentioned, if the address is not the same as exodus, I think it is either confusion between private key and seed, or a different index.

**Comment by revoxhere** [comment]
Thanks guys - that was the issue. I was never aware that XNO uses separate indexes and private keys - I still have a lot to figure out, but I'll get there. Indeed the address was different than the one Exodus shows.

Now to the solution in case anyone has the same problem:
I checked a few different indexes, but none was my actual address, so I thought maybe I can fix this by sending all coins to the address shown by index 0 and that worked, but the coins don't show anymore in my Exodus wallet (despite having the exact same private key). Funnily enough, exodus says it's only using index 0 account but it has a completely different address.

Nonetheless, I just transferred a portion of my XNO back to the Exodus address and kept a portion on nanohakase and that should be good enough for what I'm doing.

 Sorry for causing trouble and thanks for your support.
